File: vuln-sandbox/python/cwe-918/ssrf-http-get-safe.py
# SAFE: ssrf-http-get — URL validated against allowlist before making outbound request
# Rule: SsrfHttpGetUserInput | CWE-918 | Expected: TrueNegative

import requests
from urllib.parse import urlparse
from flask import Flask, request, jsonify

app = Flask(__name__)

ALLOWED_HOSTS = {'api.example.com', 'data.example.com', 'cdn.example.com'}


@app.route('/fetch')
def fetch_data():
    target_url = request.args.get('url', '')

    # SAFE: parse and validate the URL against an allowlist of trusted hosts
    try:
        parsed = urlparse(target_url)
    except Exception:
        return jsonify({'error': 'Invalid URL'}), 400

    if parsed.scheme not in ('http', 'https') or parsed.hostname not in ALLOWED_HOSTS:
        return jsonify({'error': 'URL not allowed'}), 400

    response = requests.get(target_url, timeout=5)
    return jsonify({'status': response.status_code, 'body': response.text[:1000]})


if __name__ == '__main__':
    app.run()

File: vuln-sandbox/node/cwe-918/ssrf-http-get-safe.js
// SAFE: ssrf-http-get — allowlist validation prevents SSRF
// Rule: SsrfHttpGetUserInput | CWE-918 | Expected: TrueNegative

const express = require('express');
const http = require('http');

const app = express();

const ALLOWED_HOSTS = ['api.example.com', 'data.example.com'];

app.get('/fetch', (req, res) => {
  const host = req.query.host;

  // SAFE: host validated against an allowlist before making outbound request
  if (!ALLOWED_HOSTS.includes(host)) {
    return res.status(400).json({ error: 'Host not allowed' });
  }

  http.get(`http://${host}/data`, (response) => {
    let data = '';
    response.on('data', (chunk) => { data += chunk; });
    response.on('end', () => res.send(data));
  }).on('error', (err) => res.status(500).json({ error: err.message }));
});

app.listen(3000);
